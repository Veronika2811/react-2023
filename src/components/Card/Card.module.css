.card {
  position: relative;
  width: 250px;
  text-align: center;
  list-style: none;
  border: 1px solid #4d4b4b;
  border-radius: 10px;
  transition: border-color 0.3s;
  cursor: pointer;
}

.card:hover {
  border-color: #918e8e;
}

.card__image {
  width: 100%;
  height: 250px;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
}

.card__label {
  position: absolute;
  top: 10px;
  right: 10px;
  padding: 0px 10px;
  border-radius: 5px;
}

.card__label_red {
  background-color: #ec2d01;
}

.card__label_green {
  background-color: #76c076;
}

.card__label_grey {
  background-color: #808080;
}

.card__content {
  padding: 10px;
}
