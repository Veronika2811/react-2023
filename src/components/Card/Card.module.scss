.card {
  border: 1px solid #d8d8d8;
  border-radius: 10px;
  text-align: center;
  position: relative;
  width: 250px;

  &:hover {
    cursor: pointer;
  }

  &__item {
    font-size: 15px;
  }

  &__value {
    font-style: italic;
  }

  &__image {
    height: 250px;
    width: 100%;
    border-top-left-radius: 10px;
    border-top-right-radius: 10px;
  }

  &__label {
    position: absolute;
    right: 10px;
    top: 10px;
    color: white;
    font-weight: 700;
    padding: 5px;
    border-radius: 5px;
    font-size: 15px;
  }

  &__description {
    padding: 15px;
    display: flex;
    flex-direction: column;
    gap: 5px;
  }
}
