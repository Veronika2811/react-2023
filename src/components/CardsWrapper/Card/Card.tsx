import { useRouter } from 'next/router';
import Image from 'next/image';

import { IDataResult } from '@/types/types';
import getStatusCharacterColor from '@/utils/getStatusCharacterColor/getStatusCharacterColor';
import { DEFAULT_QUERY_PARAMS } from '@/utils/constants/constants';

import classes from './Card.module.css';

const Card = ({ card }: { card: IDataResult }) => {
  const router = useRouter();

  const { id, status, image, gender, species, name, location } = card;

  const openCardDetails = () => {
    const { name: query, page, perPage } = router.query;

    router.push({
      pathname: '/',
      query: {
        details: id,
        ...(query ? { name: query } : {}),
        page: page || DEFAULT_QUERY_PARAMS.currentPage,
        perPage: perPage || DEFAULT_QUERY_PARAMS.perPage,
      },
    });
  };

  return (
    <li className={classes.card} onClick={openCardDetails} data-testid="card">
      <p
        className={`${classes.card__label} ${
          classes[getStatusCharacterColor(status)]
        }`}
      >
        {status}
      </p>
      <Image src={image} alt={name} width={250} height={250} priority={true} />
      <div className={classes.card__content}>
        <h2>{name}</h2>
        <p>{`Gender: ${gender}`}</p>
        <p>{`Species: ${species}`}</p>
        <p>{`Location: ${location.name}`}</p>
      </div>
    </li>
  );
};

export default Card;
