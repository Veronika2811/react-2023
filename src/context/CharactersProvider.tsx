import { PropsWithChildren, useState, useMemo } from 'react';
import { useSearchParams } from 'react-router-dom';

import { CharactersContext } from './context';
import {
  DEFAULT_VALUE_PER_PAGE,
  DETAILS_URL_PARAMETER_KEY,
  LOCAL_STORAGE_KEY,
} from '../constants/constants';
import { IData } from '../types/types';

export const CharactersProvider = ({ children }: PropsWithChildren) => {
  const [searchParams] = useSearchParams();

  const { Provider } = CharactersContext;

  const [searchQuery, setSearchQuery] = useState<string>(
    localStorage.getItem(LOCAL_STORAGE_KEY) || ''
  );
  const [perPage, setPerPage] = useState<number>(DEFAULT_VALUE_PER_PAGE);
  const [data, setData] = useState<IData | null>(null);
  const [detailedCard, setDetailedCard] = useState<string>(
    searchParams.get(DETAILS_URL_PARAMETER_KEY) || ''
  );

  const value = useMemo(
    () => ({
      searchQuery,
      setSearchQuery,
      perPage,
      setPerPage,
      data,
      setData,
      detailedCard,
      setDetailedCard,
    }),
    [searchQuery, perPage, data, detailedCard]
  );

  return <Provider value={value}>{children}</Provider>;
};
