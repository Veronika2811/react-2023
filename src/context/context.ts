import React, { createContext } from 'react';

import { IData } from '../types/types';

export type TCharactersContext = {
  searchQuery: string;
  setSearchQuery: React.Dispatch<React.SetStateAction<string>>;
  perPage: number;
  setPerPage: React.Dispatch<React.SetStateAction<number>>;
  data: IData | null;
  setData: React.Dispatch<React.SetStateAction<IData | null>>;
  detailedCard: string;
  setDetailedCard: React.Dispatch<React.SetStateAction<string>>;
};

export const CharactersContext = createContext<TCharactersContext>(null!);
