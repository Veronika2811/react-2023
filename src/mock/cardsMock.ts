export const MortyMock = {
  created: '2017-11-04T18:50:21.651Z',
  episode: ['https://rickandmortyapi.com/api/episode/1'],
  gender: 'Male',
  id: 2,
  image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
  location: {
    name: 'Citadel of Ricks',
    url: 'https://rickandmortyapi.com/api/location/3',
  },
  name: 'Morty Smith',
  origin: {
    name: 'unknown',
    url: '',
  },
  species: 'Human',
  status: 'Alive',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/2',
};

export const AlbertMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 11,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Albert Einstein',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};

export const RickSanchezMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 1,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Rick Sanchez',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};

export const SummerSmithMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 3,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Summer Smith',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
export const BethSmithMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 4,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Beth Smith',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};

export const JerrySmithMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 5,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Beth Smith',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};

export const AbadangoClusterPrincessMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 6,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Abadango Cluster Princess',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
export const AgencyDirectorMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 7,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Agency Director',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
export const AdjudicatorRickMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 8,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Adjudicator Rick',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
export const AbradolfLinclerMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 9,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Abradolf Lincler',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};

const dataMock = [
  MortyMock,
  AbradolfLinclerMock,
  AdjudicatorRickMock,
  AgencyDirectorMock,
  AbadangoClusterPrincessMock,
  JerrySmithMock,
  BethSmithMock,
  SummerSmithMock,
  RickSanchezMock,
  AlbertMock,
];

export const initialCharacterArrayMock = [...dataMock, ...dataMock];

export const first10СharactersMock = initialCharacterArrayMock.slice(0, 10);
export const last10СharactersMock = initialCharacterArrayMock.slice(0, -10);
