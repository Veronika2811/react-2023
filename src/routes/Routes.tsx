import { createHashRouter } from 'react-router-dom';

import App from '../components/App/App';
import NothingFound from '../components/NothingFound/NothingFound';
import CardDetails from '../components/CardDetails/CardDetails';

const Routes = createHashRouter([
  {
    path: '/',
    element: <App />,
    errorElement: <NothingFound error="error-router" />,
    children: [
      {
        path: '',
        element: <CardDetails />,
      },
    ],
  },
]);

export default Routes;
