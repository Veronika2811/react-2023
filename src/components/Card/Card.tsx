import { useContext } from 'react';
import { useSearchParams } from 'react-router-dom';

import getStatusCharacterColor from '../../utils/getStatusCharacterColor';
import { CharactersContext } from '../../context/context';
import { IDataResult } from '../../types/types';
import { DETAILS_URL_PARAMETER_KEY } from '../../constants/constants';

import classes from './Card.module.css';

interface ICharacterCardProps {
  card: IDataResult;
}

const Card = ({ card }: ICharacterCardProps) => {
  const [, setSearchParams] = useSearchParams();
  const { setDetailedCard } = useContext(CharactersContext);

  const { id, status, image, name, gender, species, location } = card;

  return (
    <li
      className={classes.card}
      onClick={() => {
        const cardId = id.toString();
        setSearchParams((searchParams) => {
          searchParams.set(DETAILS_URL_PARAMETER_KEY, cardId);
          return searchParams;
        });
        setDetailedCard(cardId);
      }}
      data-testid="card"
    >
      <p
        className={`${classes.card__label} ${
          classes[getStatusCharacterColor(status)]
        }`}
      >
        {status}
      </p>
      <img
        data-testid="card-image"
        className={classes.card__image}
        src={image}
        alt={name}
      />
      <div className={classes.card__content}>
        <h2>{name}</h2>
        <p>{`Gender: ${gender}`}</p>
        <p>{`Species: ${species}`}</p>
        <p>{`Location: ${location.name}`}</p>
      </div>
    </li>
  );
};

export default Card;
