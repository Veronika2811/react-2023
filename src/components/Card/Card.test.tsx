import { expect, it } from 'vitest';
import { HashRouter, RouterProvider } from 'react-router-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { CharactersProvider } from '../../context/CharactersProvider';
import Card from './Card';
import Routes from '../../routes/Routes';
import { MortyMock } from '../../mock/cardsMock';

describe('Card component', () => {
  it('renders correctly Card component', () => {
    const container = render(
      <HashRouter>
        <CharactersProvider>
          <Card card={MortyMock} />
        </CharactersProvider>
      </HashRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('should display relevant character data Card component', () => {
    render(
      <HashRouter>
        <CharactersProvider>
          <Card card={MortyMock} />
        </CharactersProvider>
      </HashRouter>
    );

    const { status, image, name, gender, species, location } = MortyMock;

    const STATUS = screen.getByText(status);
    expect(STATUS).toBeInTheDocument();

    const CARD_IMAGE = screen.queryByTestId('card-image') as HTMLImageElement;
    expect(CARD_IMAGE.src).toContain(image);
    expect(CARD_IMAGE.alt).toContain(name);

    const NAME = screen.getByText(name);
    expect(NAME).toBeInTheDocument();

    const GENDER = screen.getByText(`Gender: ${gender}`);
    expect(GENDER).toBeInTheDocument();

    const SPECIES = screen.getByText(`Species: ${species}`);
    expect(SPECIES).toBeInTheDocument();

    const LOCATION_NAME = screen.getByText(`Location: ${location.name}`);
    expect(LOCATION_NAME).toBeInTheDocument();
  });

  it('should open detailed map component when clicking on map', async () => {
    render(<RouterProvider router={Routes} />);

    const cardDetails = screen.queryByTestId('card-details');

    expect(cardDetails).not.toBeInTheDocument();

    await waitFor(async () =>
      fireEvent.click(screen.getAllByTestId('card')[1])
    );

    await waitFor(() => {
      const cardDetails = screen.getByTestId('card-details');
      expect(cardDetails).toBeInTheDocument();
    });
  });
});
