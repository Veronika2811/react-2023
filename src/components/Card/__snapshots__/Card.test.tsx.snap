// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > renders correctly Card component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="_card_1f2933"
        data-testid="card"
      >
        <p
          class="_card__label_1f2933 _card__label_green_1f2933"
        >
          Alive
        </p>
        <img
          alt="Morty Smith"
          class="_card__image_1f2933"
          data-testid="card-image"
          src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
        />
        <div
          class="_card__content_1f2933"
        >
          <h2>
            Morty Smith
          </h2>
          <p>
            Gender: Male
          </p>
          <p>
            Species: Human
          </p>
          <p>
            Location: Citadel of Ricks
          </p>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="_card_1f2933"
      data-testid="card"
    >
      <p
        class="_card__label_1f2933 _card__label_green_1f2933"
      >
        Alive
      </p>
      <img
        alt="Morty Smith"
        class="_card__image_1f2933"
        data-testid="card-image"
        src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
      />
      <div
        class="_card__content_1f2933"
      >
        <h2>
          Morty Smith
        </h2>
        <p>
          Gender: Male
        </p>
        <p>
          Species: Human
        </p>
        <p>
          Location: Citadel of Ricks
        </p>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
